rust/storage/src/admissioncontrolleds3.rs

```diff
@@ -57,22 +61,74 @@ struct InflightRequest {
     >,
 }
 
+type InflightRequestFuture = Shared<
+    Pin<
+        Box<
+            dyn Future<Output = Result<(Arc<Vec<u8>>, Option<ETag>), StorageError>>
+                + Send
+                + 'static,
+        >,
+    >,
+>;
+
 impl InflightRequest {
+    pub fn new(
+        priority_holder: Arc<PriorityHolder>,
+        future: InflightRequestFuture,
+        notify_channel: Option<tokio::sync::mpsc::Sender<()>>,
+    ) -> Self {
+        Self {
+            finished: Arc::new(AtomicBool::new(false)),
+            priority_holder,
+            notify_channel,
+            future,
+        }
+    }
+
     // Not thread safe.
-    async fn update_priority(&self, priority: StorageRequestPriority) {
+    async fn update_priority(&self, priority: StorageRequestPriority) -> RollbackPriorityOnDrop {
         // It is ok to not do Compare And Swap here since the caller obtains a mutex before
         // performing this operation so at any point there will only be one writer
         // for this AtomicUsize.
+        let curr_pri = self.priority_holder.get_priority();
+        self.priority_holder.add_priority(priority);
+
+        let guard = RollbackPriorityOnDrop(self.clone(), priority);
         if let Some(channel) = &self.notify_channel {
-            let curr_pri = self.priority.load(std::sync::atomic::Ordering::SeqCst);
-            if priority.as_usize() < curr_pri {
-                self.priority
-                    .store(priority.as_usize(), std::sync::atomic::Ordering::SeqCst);
+            if priority.as_usize() != curr_pri.as_usize() {
```

Comment:
Since we are now calling `update_priority` even from the primary driver of the request - can this result in it sending a request to itself? Probably an optimization to skip if `curr_pri` is uninitialized?

Reply:
Also, why do you need a `!=` here? Shouldn't this continue to be `<`? You send a new notification on the drop which should take care of decreasing the priority no? And during the add phase, notifications should only be sent for priority increases?